route/v1/types.go

```diff
@@ -139,8 +139,90 @@ type RouteSpec struct {
 	// +kubebuilder:default=None
 	WildcardPolicy WildcardPolicyType `json:"wildcardPolicy,omitempty" protobuf:"bytes,7,opt,name=wildcardPolicy"`
+
+	// httpHeaders defines policy for HTTP headers.
+	//
+	// If this field is empty, the default values are used.
+	//
+	// +optional
+	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`
+}
+
+// RouteHTTPHeaders defines policy for HTTP headers.
+type RouteHTTPHeaders struct {
+	// actions specifies options for performing action on headers.
+	// For Setting Headers one can provide a value for the header whereas for deleting one needs to provide header name.
+	// +optional
+	Actions RouteHTTPHeaderActions `json:"actions,omitempty" protobuf:"bytes,1,opt,name=actions"`
+}
+
+// RouteHTTPHeaderActions defines whether we want to set Response or Request Headers.
+type RouteHTTPHeaderActions struct {
+	// Response is a list of http response headers which either need to set or deleted.
+	Response []RouteHTTPHeader `json:"response,omitempty" protobuf:"bytes,1,opt,name=response"`
+	// Request is a list of http request headers which either need to set or deleted.
+	Request []RouteHTTPHeader `json:"request,omitempty" protobuf:"bytes,2,opt,name=request"`
+}
+
+// RouteHTTPHeader specifies options for set/delete a header.
+type RouteHTTPHeader struct {
+	// name specifies a header name to be set or deleted.  Its value must be a valid HTTP header.
+	// name as defined in RFC 2616 section 4.2.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	// +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// value specifies fields to mention the header details and the action you want to perform on them.
+	Value RouteHTTPHeaderActionUnion `json:"value" protobuf:"bytes,2,opt,name=value"`
 }
 
+// RouteHTTPHeaderActionUnion specifies fields to mention the header details and the action you want to perform on them.
+// +union
+type RouteHTTPHeaderActionUnion struct {
+	// action defines the type of the action to be applied on the header.
+	// Set allows you to set http request and response headers.
+	// Delete allows you to delete http request and response headers
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=Set;Delete
+	// +required
+	Action RouteHTTPHeaderActionType `json:"action,omitempty" protobuf:"bytes,11,opt,name=action,casttype=RouteHTTPHeaderActionType"`
+
+	// set defines the HTTP header that should be set: added if doesn't exist or replaced if does.
+	// If this field is empty, no header is added or replaced.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be set for TLS passthrough
+	// connections.
```

Comment:
This note belongs on `spec.httpHeaders`, and it can be contextualized for the Route API:
```suggestion
	// Note that actions on HTTP headers can only be specified if spec.tls
	// is not specified or if spec.tls.termination is set to "edge" or
	// "reencrypt".  Actions cannot be specified for TLS passthrough routes.
```
Will validation enforce this restriction?

Reply:
no. we can check this in library-go and admission code

Reply:
We could use CEL validation to enforce this condition for the CRD version of this API, why wouldn't we do that?
I'm guessing CEL cannot be used for extension API servers such as openshift-api?

Reply:
Actually i have already added a validation in library go code. I am not sure if CEL will work.

Reply:
You can add CEL validation for this too, I'd like to see CEL matching the validations you are doing in library go as this API is served via a CRD in MicroShift IIUC